{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "e5db7d67",
   "metadata": {},
   "source": [
    "# Project 1 - Team 2\n",
    "\n",
    "## Xiaotao Dong, Kelly Mayhew, Exequiel Punzalan, Di Wang\n",
    "\n",
    "### Project 1 - Drug annotation of 23andme report (instructions copied from Canvas, delete later)\n",
    "\n",
    "23andme provides the raw data regarding variants in a tab separated file.\n",
    "\n",
    "Examples of 23andme files are available at:\n",
    "\n",
    "https://github.com/arrogantrobot/23andme2vcf (Links to an external site.)\n",
    "\n",
    " \n",
    "For this project use: 23andme_v5_hg19_ref.txt.gz from the above link.\n",
    "\n",
    "The columns in this file are the following:\n",
    "\n",
    "`CHR, POS, dbSNP_ID, ALLELE`\n",
    " \n",
    "\n",
    "Details about the format can be found at:\n",
    "\n",
    "https://customercare.23andme.com/hc/en-us/articles/115004459928-Raw-Genotype-Data-Technical-Details (Links to an external site.)\n",
    "\n",
    "https://samtools.github.io/bcftools/howtos/convert.html (Links to an external site.)\n",
    " \n",
    "\n",
    "dbSNP_ID is the variant identifier in the database dbSNP (https://www.ncbi.nlm.nih.gov/snp/ (Links to an external site.)) when is available, or . , when is unavailable\n",
    " \n",
    "\n",
    "Based on the 23andme file, __detect variants with drug effects as provided by **PharmGKB**.__\n",
    "\n",
    "PharmGKB data is available at:\n",
    "\n",
    "https://www.pharmgkb.org/downloads (Links to an external site.)\n",
    "\n",
    " \n",
    "\n",
    "To connect the variants to the information available in the PharmGKB  data we need the **var_drug_ann.tsv** available at the above link in the **annotations.zip** archive.\n",
    "\n",
    " \n",
    "\n",
    "The **var_drug_ann.tsv** file contains the following information in a tab separated format:\n",
    "\n",
    "(Variant file description at: https://www.pharmgkb.org/page/downloadVariantAnnotationsHelp)\n",
    "\n",
    "(The file has a bad line 8140, handle the bad line (either fix it, manually, or skip it programatically, but only skip that line).\n",
    "\n",
    " \n",
    "```\n",
    "Annotation ID: unique ID number for each variant/drug annotation\n",
    "\n",
    "Variant: dbSNP ID or haplotype\n",
    "\n",
    "Gene: HGNC symbol (PharmGKB Accession ID)\n",
    "\n",
    "Chemical: Drug name (PharmGKB Accession ID)\n",
    "\n",
    "Literature ID: PMID - this is PubMed identifier\n",
    "\n",
    "Phenotype Category: options [efficacy, toxicity, dosage, metabolism/PK other]\n",
    "\n",
    "Significance: yes or no – determined by if the author stated the association was significant\n",
    "\n",
    "Notes: curator notes field\n",
    "\n",
    "Sentence: structured sentence\n",
    "\n",
    "StudyParameters: corresponds with the Study Parameters ID in the study_parameters.tsv file\n",
    "\n",
    "Alleles: variant alleles in annotation\n",
    "```\n",
    " \n",
    "\n",
    "1. Map/merge the 23andme file and the variant-drug annotation file based on dbSNP_ID (also known as rsID).\n",
    "\n",
    "The merged result should have the following columns:\n",
    "\n",
    "`dbSNP_ID, GENE_SYMBOL, DRUG_NAME, PMID, PHENOTYPE_CATEGORY, SIGNIFICANCE, NOTES, SENTENCE, ALLELE_PharmGKB (variant alleles in annotation), ALLELE_23andme (variant alleles in 23andme file)`\n",
    "\n",
    "\n",
    "2. Filter the output so that it only contains significant associations (SIGNIFICANCE is yes) for variants that affect the drug efficacy ('PHENOTYPE_CATEGORY' is 'efficacy').\n",
    "\n",
    " \n",
    "\n",
    "3. Save the output of the filtering step in a tab-separated file (23andme_PharmGKB_map.tsv) with the following columns:\n",
    "\n",
    "`dbSNP_ID, GENE_SYMBOL, DRUG_NAME, NOTES, SENTENCE, ALLELE_PharmGKB, ALLELE_23andme`\n",
    "\n",
    " \n",
    "\n",
    "4. Create a tab separated file with summarized data with the following columns:\n",
    "\n",
    "`GENE_SYMBOL, DRUG_NAME, and a list of dbSNP_IDs separated by \";\"`\n",
    "\n",
    " \n",
    "\n",
    "5. Plot the distribution of the number of drugs associated with a gene, and the number of SNPs for a gene. \n",
    "\n",
    "---------------\n",
    "\n",
    "Things to consider when designing and writing your solution:\n",
    "\n",
    "* break the code in small parts\n",
    "  * it can help you distribute the work among the group members\n",
    "  * it makes the code writing easier: the small parts are easier to understand and implement\n",
    "* make it self contained\n",
    "  * if any additional modules need to be imported/installed provide that information in the solution\n",
    "* document/comment your code\n",
    "  * provide instructions of how to run you code\n",
    "    * This can be provided in in Markdown cells towards the beginning of the notebook (after the answer to the questions)\n",
    "    * Or as docstrings towards the beginning of the script file (after the answer to the questions)\n",
    "* do not hardcode if it can be avoided\n",
    "* do not write redundant code\n",
    "* write functions and classes when applicable\n",
    "* keep it simple, do not to overly complicate the solution\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f19e7be9",
   "metadata": {},
   "source": [
    "One part of the solution is the code the team will write to compute the required output from the provided input.\n",
    "\n",
    "In addition, each team member has to provide an answer to the following three questions:\n",
    "1. What was your biggest challenge in this project? (regarding writing code and not only)\n",
    "2. What did you learn while working on this project? (regarding writing code and not only)\n",
    "3. If you had more time on the project what other question(s) would you like to answer? (at least one question is required)\n",
    "\n",
    "The answers written by each team member should be added in Markdown cells at the beginning of the notebook, or as docstrings at the beginning of the script file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76c367fa",
   "metadata": {},
   "source": [
    "## Questions - Xiaotao Dong"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b76ef7a",
   "metadata": {},
   "source": [
    "## Questions - Kelly Mayhew"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f44d3af",
   "metadata": {},
   "source": [
    "## Questions - Exequiel Punzalan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24ed73cf",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "cc48ab99",
   "metadata": {},
   "source": [
    "## Questions - Di Wang"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89e4e4ee",
   "metadata": {},
   "source": [
    "## Problem Solution"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01c88bd7",
   "metadata": {},
   "source": [
    "### Part 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "8833f498",
   "metadata": {},
   "outputs": [],
   "source": [
    "#1. Map/merge the 23andme file and the variant-drug annotation file based on dbSNP_ID (also known as rsID).\n",
    "# The merged result should have the following columns:\n",
    "# dbSNP_ID, GENE_SYMBOL, DRUG_NAME, PMID, PHENOTYPE_CATEGORY, SIGNIFICANCE, NOTES, \n",
    "# SENTENCE, ALLELE_PharmGKB (variant alleles in annotation), ALLELE_23andme (variant alleles in 23andme file)\n",
    "\n",
    "variant_file = \"23andme_v5_hg19_ref.txt.gz\"\n",
    "annotation_file = \"var_drug_ann.tsv\"\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import gzip\n",
    "def map_and_merge(gene_variants, drug_annotations):\n",
    "    \"\"\"\n",
    "    Read in two files (both tab-separated, one .tsv and one .txt.gz), and combine them into a pandas dataframe.\n",
    "    \n",
    "    @param gene_variants: a .txt.gz file of gene variants from 23andme\n",
    "    @param drug_annotations: a .tsv file of drug annotations\n",
    "    @return a pandas dataframe of merged gene variants and drug annotations\n",
    "    \"\"\"\n",
    "    \n",
    "    #Create gene variants dataframe\n",
    "    genecol=[\"CHR\",\"POS\",\"dbSNP_ID\",\"ALLELE_23andme\"]\n",
    "    with gzip.open(gene_variants) as file:\n",
    "        genes=pd.read_csv(file, sep=\"\\t\", names=genecol)\n",
    "    \n",
    "    #Create drug annotations dataframe\n",
    "    drugcol = [\"ANNOTATION_ID\",\"dbSNP_ID\",\"GENE_SYMBOL\",\"DRUG_NAME\",\"PMID\",\"PHENOTYPE_CATEGORY\",\"SIGNIFICANCE\",\\\n",
    "              \"NOTES\",\"SENTENCE\",\"ALLELE_PharmKGB\",\"STUDY_PARAMETERS\"]\n",
    "    drugs = pd.read_csv(drug_annotations, sep=\"\\t\", names=drugcol, skiprows=1, error_bad_lines=False)\n",
    "    \n",
    "    #Join results\n",
    "    merged_table = pd.merge(drugs, genes, on=\"dbSNP_ID\")\n",
    "    #Drop unwanted columns\n",
    "    merged_table = merged_table.drop([\"CHR\",\"POS\",\"ANNOTATION_ID\",\"STUDY_PARAMETERS\"], axis=1)\n",
    "    \n",
    "    return merged_table\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ba63986a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbSNP_ID</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>PMID</th>\n",
       "      <th>PHENOTYPE_CATEGORY</th>\n",
       "      <th>SIGNIFICANCE</th>\n",
       "      <th>NOTES</th>\n",
       "      <th>SENTENCE</th>\n",
       "      <th>ALLELE_PharmKGB</th>\n",
       "      <th>ALLELE_23andme</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs5031016</td>\n",
       "      <td>CYP2A6</td>\n",
       "      <td>warfarin</td>\n",
       "      <td>22248286</td>\n",
       "      <td>Dosage</td>\n",
       "      <td>no</td>\n",
       "      <td>No association was found between this variant ...</td>\n",
       "      <td>Allele G is not associated with increased dose...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs5031016</td>\n",
       "      <td>CYP2A6</td>\n",
       "      <td>nicotine</td>\n",
       "      <td>11779172</td>\n",
       "      <td>Metabolism/PK</td>\n",
       "      <td>not stated</td>\n",
       "      <td>This SNP is in the CYP2A6*10 and *7 alleles ex...</td>\n",
       "      <td>Allele G is associated with decreased metaboli...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs5031016</td>\n",
       "      <td>CYP2A6</td>\n",
       "      <td>nicotine</td>\n",
       "      <td>12445030</td>\n",
       "      <td>Other,\"Metabolism/PK\"</td>\n",
       "      <td>no</td>\n",
       "      <td>This is a SNP within the CYP2A6*7 or *10 allel...</td>\n",
       "      <td>Allele G is associated with decreased metaboli...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs7624046</td>\n",
       "      <td>KCNMB2</td>\n",
       "      <td>ritodrine</td>\n",
       "      <td>32371615</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>variant-type homozygote carriers of rs7624046 ...</td>\n",
       "      <td>Genotype TT is associated with decreased respo...</td>\n",
       "      <td>TT</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs1128503</td>\n",
       "      <td>ABCB1</td>\n",
       "      <td>fentanyl</td>\n",
       "      <td>29601950</td>\n",
       "      <td>Dosage</td>\n",
       "      <td>yes</td>\n",
       "      <td>Please note that alleles have been complemente...</td>\n",
       "      <td>Genotype AA is associated with increased dose ...</td>\n",
       "      <td>AA</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5277</th>\n",
       "      <td>rs496503</td>\n",
       "      <td>KCNT1</td>\n",
       "      <td>antiepileptics</td>\n",
       "      <td>24279416</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>no</td>\n",
       "      <td>No significant differences in genotype frequen...</td>\n",
       "      <td>Allele A is not associated with response to an...</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5278</th>\n",
       "      <td>rs1948308</td>\n",
       "      <td>NTRK2</td>\n",
       "      <td>methadone</td>\n",
       "      <td>23651024</td>\n",
       "      <td>Dosage</td>\n",
       "      <td>no</td>\n",
       "      <td>This association was statistically significant...</td>\n",
       "      <td>Genotype GG is associated with decreased dose ...</td>\n",
       "      <td>GG</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5279</th>\n",
       "      <td>rs4742610</td>\n",
       "      <td>PTPRD</td>\n",
       "      <td>trandolapril,\"verapamil\"</td>\n",
       "      <td>26425837</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>This was associated with resistant hypertensio...</td>\n",
       "      <td>Allele T is associated with resistance to tran...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5280</th>\n",
       "      <td>rs1056515</td>\n",
       "      <td>RGS5</td>\n",
       "      <td>bevacizumab,\"fluorouracil\",\"irinotecan\",\"leuco...</td>\n",
       "      <td>25069475</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>For specifically tumors occurring in the right...</td>\n",
       "      <td>Genotype TT is associated with increased respo...</td>\n",
       "      <td>TT</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5281</th>\n",
       "      <td>rs3093158</td>\n",
       "      <td>CYP4F2</td>\n",
       "      <td>warfarin</td>\n",
       "      <td>25126975</td>\n",
       "      <td>Dosage</td>\n",
       "      <td>no</td>\n",
       "      <td>Samples, genotypes and INRs from a cohort of 5...</td>\n",
       "      <td>Allele T is not associated with dose of warfar...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5282 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       dbSNP_ID GENE_SYMBOL  \\\n",
       "0     rs5031016      CYP2A6   \n",
       "1     rs5031016      CYP2A6   \n",
       "2     rs5031016      CYP2A6   \n",
       "3     rs7624046      KCNMB2   \n",
       "4     rs1128503       ABCB1   \n",
       "...         ...         ...   \n",
       "5277   rs496503       KCNT1   \n",
       "5278  rs1948308       NTRK2   \n",
       "5279  rs4742610       PTPRD   \n",
       "5280  rs1056515        RGS5   \n",
       "5281  rs3093158      CYP4F2   \n",
       "\n",
       "                                              DRUG_NAME      PMID  \\\n",
       "0                                              warfarin  22248286   \n",
       "1                                              nicotine  11779172   \n",
       "2                                              nicotine  12445030   \n",
       "3                                             ritodrine  32371615   \n",
       "4                                              fentanyl  29601950   \n",
       "...                                                 ...       ...   \n",
       "5277                                     antiepileptics  24279416   \n",
       "5278                                          methadone  23651024   \n",
       "5279                           trandolapril,\"verapamil\"  26425837   \n",
       "5280  bevacizumab,\"fluorouracil\",\"irinotecan\",\"leuco...  25069475   \n",
       "5281                                           warfarin  25126975   \n",
       "\n",
       "         PHENOTYPE_CATEGORY SIGNIFICANCE  \\\n",
       "0                    Dosage           no   \n",
       "1             Metabolism/PK   not stated   \n",
       "2     Other,\"Metabolism/PK\"           no   \n",
       "3                  Efficacy          yes   \n",
       "4                    Dosage          yes   \n",
       "...                     ...          ...   \n",
       "5277               Efficacy           no   \n",
       "5278                 Dosage           no   \n",
       "5279               Efficacy          yes   \n",
       "5280               Efficacy          yes   \n",
       "5281                 Dosage           no   \n",
       "\n",
       "                                                  NOTES  \\\n",
       "0     No association was found between this variant ...   \n",
       "1     This SNP is in the CYP2A6*10 and *7 alleles ex...   \n",
       "2     This is a SNP within the CYP2A6*7 or *10 allel...   \n",
       "3     variant-type homozygote carriers of rs7624046 ...   \n",
       "4     Please note that alleles have been complemente...   \n",
       "...                                                 ...   \n",
       "5277  No significant differences in genotype frequen...   \n",
       "5278  This association was statistically significant...   \n",
       "5279  This was associated with resistant hypertensio...   \n",
       "5280  For specifically tumors occurring in the right...   \n",
       "5281  Samples, genotypes and INRs from a cohort of 5...   \n",
       "\n",
       "                                               SENTENCE ALLELE_PharmKGB  \\\n",
       "0     Allele G is not associated with increased dose...               G   \n",
       "1     Allele G is associated with decreased metaboli...               G   \n",
       "2     Allele G is associated with decreased metaboli...               G   \n",
       "3     Genotype TT is associated with decreased respo...              TT   \n",
       "4     Genotype AA is associated with increased dose ...              AA   \n",
       "...                                                 ...             ...   \n",
       "5277  Allele A is not associated with response to an...               A   \n",
       "5278  Genotype GG is associated with decreased dose ...              GG   \n",
       "5279  Allele T is associated with resistance to tran...               T   \n",
       "5280  Genotype TT is associated with increased respo...              TT   \n",
       "5281  Allele T is not associated with dose of warfar...               T   \n",
       "\n",
       "     ALLELE_23andme  \n",
       "0                 A  \n",
       "1                 A  \n",
       "2                 A  \n",
       "3                 C  \n",
       "4                 A  \n",
       "...             ...  \n",
       "5277              G  \n",
       "5278              G  \n",
       "5279              C  \n",
       "5280              G  \n",
       "5281              C  \n",
       "\n",
       "[5282 rows x 10 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TEST PART 1\n",
    "\n",
    "map_and_merge(variant_file, annotation_file)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a7a44f1",
   "metadata": {},
   "source": [
    "### Part 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "1b79545d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbSNP_ID</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>PMID</th>\n",
       "      <th>PHENOTYPE_CATEGORY</th>\n",
       "      <th>SIGNIFICANCE</th>\n",
       "      <th>NOTES</th>\n",
       "      <th>SENTENCE</th>\n",
       "      <th>ALLELE_PharmKGB</th>\n",
       "      <th>ALLELE_23andme</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs7624046</td>\n",
       "      <td>KCNMB2</td>\n",
       "      <td>ritodrine</td>\n",
       "      <td>32371615</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>variant-type homozygote carriers of rs7624046 ...</td>\n",
       "      <td>Genotype TT is associated with decreased respo...</td>\n",
       "      <td>TT</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>rs1128503</td>\n",
       "      <td>ABCB1</td>\n",
       "      <td>fentanyl</td>\n",
       "      <td>22277678</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>Please note that alleles have been complemente...</td>\n",
       "      <td>Genotype AA is associated with decreased respo...</td>\n",
       "      <td>AA</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>rs1128503</td>\n",
       "      <td>ABCB1</td>\n",
       "      <td>glucocorticoids</td>\n",
       "      <td>25346426</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>The GG genotype was more frequent in patients ...</td>\n",
       "      <td>Genotype GG is associated with decreased respo...</td>\n",
       "      <td>GG</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>rs1128503</td>\n",
       "      <td>ABCB1</td>\n",
       "      <td>antipsychotics</td>\n",
       "      <td>22909202</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>Please note: this variant deviated from Hardy-...</td>\n",
       "      <td>Allele G is associated with increased response...</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>rs1128503</td>\n",
       "      <td>ABCB1</td>\n",
       "      <td>tacrolimus</td>\n",
       "      <td>29615122</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>Patients with the AA genotype had a higher eff...</td>\n",
       "      <td>Genotype AA is associated with increased respo...</td>\n",
       "      <td>AA</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5265</th>\n",
       "      <td>rs4796793</td>\n",
       "      <td>STAT3</td>\n",
       "      <td>interferons</td>\n",
       "      <td>17602083</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>IFN-alpha. Careful- GC SNP. Based on dbSNP pop...</td>\n",
       "      <td>Genotype GG is associated with increased respo...</td>\n",
       "      <td>GG</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5272</th>\n",
       "      <td>rs2233945</td>\n",
       "      <td>PSORS1C1</td>\n",
       "      <td>etanercept</td>\n",
       "      <td>28107378</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>The degree of response was determined by EULAR...</td>\n",
       "      <td>Allele C is associated with decreased response...</td>\n",
       "      <td>C</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5274</th>\n",
       "      <td>rs2900478</td>\n",
       "      <td>SLCO1B1</td>\n",
       "      <td>hmg coa reductase inhibitors</td>\n",
       "      <td>25350695</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>Carriers of this variant respond to statins wi...</td>\n",
       "      <td>Allele A is associated with decreased response...</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5279</th>\n",
       "      <td>rs4742610</td>\n",
       "      <td>PTPRD</td>\n",
       "      <td>trandolapril,\"verapamil\"</td>\n",
       "      <td>26425837</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>This was associated with resistant hypertensio...</td>\n",
       "      <td>Allele T is associated with resistance to tran...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5280</th>\n",
       "      <td>rs1056515</td>\n",
       "      <td>RGS5</td>\n",
       "      <td>bevacizumab,\"fluorouracil\",\"irinotecan\",\"leuco...</td>\n",
       "      <td>25069475</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>For specifically tumors occurring in the right...</td>\n",
       "      <td>Genotype TT is associated with increased respo...</td>\n",
       "      <td>TT</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1239 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       dbSNP_ID GENE_SYMBOL  \\\n",
       "3     rs7624046      KCNMB2   \n",
       "8     rs1128503       ABCB1   \n",
       "9     rs1128503       ABCB1   \n",
       "32    rs1128503       ABCB1   \n",
       "35    rs1128503       ABCB1   \n",
       "...         ...         ...   \n",
       "5265  rs4796793       STAT3   \n",
       "5272  rs2233945    PSORS1C1   \n",
       "5274  rs2900478     SLCO1B1   \n",
       "5279  rs4742610       PTPRD   \n",
       "5280  rs1056515        RGS5   \n",
       "\n",
       "                                              DRUG_NAME      PMID  \\\n",
       "3                                             ritodrine  32371615   \n",
       "8                                              fentanyl  22277678   \n",
       "9                                       glucocorticoids  25346426   \n",
       "32                                       antipsychotics  22909202   \n",
       "35                                           tacrolimus  29615122   \n",
       "...                                                 ...       ...   \n",
       "5265                                        interferons  17602083   \n",
       "5272                                         etanercept  28107378   \n",
       "5274                       hmg coa reductase inhibitors  25350695   \n",
       "5279                           trandolapril,\"verapamil\"  26425837   \n",
       "5280  bevacizumab,\"fluorouracil\",\"irinotecan\",\"leuco...  25069475   \n",
       "\n",
       "     PHENOTYPE_CATEGORY SIGNIFICANCE  \\\n",
       "3              Efficacy          yes   \n",
       "8              Efficacy          yes   \n",
       "9              Efficacy          yes   \n",
       "32             Efficacy          yes   \n",
       "35             Efficacy          yes   \n",
       "...                 ...          ...   \n",
       "5265           Efficacy          yes   \n",
       "5272           Efficacy          yes   \n",
       "5274           Efficacy          yes   \n",
       "5279           Efficacy          yes   \n",
       "5280           Efficacy          yes   \n",
       "\n",
       "                                                  NOTES  \\\n",
       "3     variant-type homozygote carriers of rs7624046 ...   \n",
       "8     Please note that alleles have been complemente...   \n",
       "9     The GG genotype was more frequent in patients ...   \n",
       "32    Please note: this variant deviated from Hardy-...   \n",
       "35    Patients with the AA genotype had a higher eff...   \n",
       "...                                                 ...   \n",
       "5265  IFN-alpha. Careful- GC SNP. Based on dbSNP pop...   \n",
       "5272  The degree of response was determined by EULAR...   \n",
       "5274  Carriers of this variant respond to statins wi...   \n",
       "5279  This was associated with resistant hypertensio...   \n",
       "5280  For specifically tumors occurring in the right...   \n",
       "\n",
       "                                               SENTENCE ALLELE_PharmKGB  \\\n",
       "3     Genotype TT is associated with decreased respo...              TT   \n",
       "8     Genotype AA is associated with decreased respo...              AA   \n",
       "9     Genotype GG is associated with decreased respo...              GG   \n",
       "32    Allele G is associated with increased response...               G   \n",
       "35    Genotype AA is associated with increased respo...              AA   \n",
       "...                                                 ...             ...   \n",
       "5265  Genotype GG is associated with increased respo...              GG   \n",
       "5272  Allele C is associated with decreased response...               C   \n",
       "5274  Allele A is associated with decreased response...               A   \n",
       "5279  Allele T is associated with resistance to tran...               T   \n",
       "5280  Genotype TT is associated with increased respo...              TT   \n",
       "\n",
       "     ALLELE_23andme  \n",
       "3                 C  \n",
       "8                 A  \n",
       "9                 A  \n",
       "32                A  \n",
       "35                A  \n",
       "...             ...  \n",
       "5265              G  \n",
       "5272              C  \n",
       "5274              T  \n",
       "5279              C  \n",
       "5280              G  \n",
       "\n",
       "[1239 rows x 10 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#2. Filter the output so that it only contains significant associations (SIGNIFICANCE is yes) for variants\n",
    "# that affect the drug efficacy ('PHENOTYPE_CATEGORY' is 'Efficacy').\n",
    "\n",
    "data = map_and_merge(variant_file, annotation_file)\n",
    "data[(data.SIGNIFICANCE == \"yes\") & (data.PHENOTYPE_CATEGORY == \"Efficacy\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "53cc9b86",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TEST PART 2\n",
    "\n",
    "testDat2 = pd.DataFrame([['rs5031016','CYP2A6','warfarin','22248286','Dosage','no','No association...','Allele G is not...'\\\n",
    "                         'G','A'],['rs496503','KCNT1','antiepileptics','24279416','Efficacy','no','No significant...'\\\n",
    "                        'Allele A is not...','A','G'],['rs1056515','RGS5','bevacizumab...','25069475','Efficacy','yes',\\\n",
    "                        'For specifically...','Genotype TT is...','TT','G']],\\\n",
    "                       columns=['dbSNP_ID','GENE_SYMBOL','DRUG_NAME','PMID','PHENOTYPE_CATEGORY','SIGNIFICANCE','NOTES',\\\n",
    "                               'SENTENCE','ALLELE_PharmKGB','ALLELE_23andme'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "258a8235-4655-4d99-9384-88cf77c10e94",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dbSNP_ID</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>DRUG_NAME</th>\n",
       "      <th>PMID</th>\n",
       "      <th>PHENOTYPE_CATEGORY</th>\n",
       "      <th>SIGNIFICANCE</th>\n",
       "      <th>NOTES</th>\n",
       "      <th>SENTENCE</th>\n",
       "      <th>ALLELE_PharmKGB</th>\n",
       "      <th>ALLELE_23andme</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs1056515</td>\n",
       "      <td>RGS5</td>\n",
       "      <td>bevacizumab...</td>\n",
       "      <td>25069475</td>\n",
       "      <td>Efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>For specifically...</td>\n",
       "      <td>Genotype TT is...</td>\n",
       "      <td>TT</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    dbSNP_ID GENE_SYMBOL       DRUG_NAME      PMID PHENOTYPE_CATEGORY  \\\n",
       "2  rs1056515        RGS5  bevacizumab...  25069475           Efficacy   \n",
       "\n",
       "  SIGNIFICANCE                NOTES           SENTENCE ALLELE_PharmKGB  \\\n",
       "2          yes  For specifically...  Genotype TT is...              TT   \n",
       "\n",
       "  ALLELE_23andme  \n",
       "2              G  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "testDat2[(testDat2.SIGNIFICANCE == \"yes\") & (testDat2.PHENOTYPE_CATEGORY == \"Efficacy\")]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b74a889",
   "metadata": {},
   "source": [
    "### Part 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1875d48",
   "metadata": {},
   "outputs": [],
   "source": [
    "#3. Save the output of the filtering step in a tab-separated file (23andme_PharmGKB_map.tsv) with the following columns:\n",
    "# dbSNP_ID, GENE_SYMBOL, DRUG_NAME, NOTES, SENTENCE, ALLELE_PharmGKB, ALLELE_23andme\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aab9b637",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TEST PART 3\n",
    "\n",
    "testDat3 = pd.DataFrame([['rs1056515','RGS5','bevacizumab...','25069475','Efficacy','yes',\\\n",
    "                        'For specifically...','Genotype TT is...','TT','G'],['rs5031016','CYP2A6','warfarin','22248286',\\\n",
    "                        'Efficacy','yes','No association...','Allele G is not...','G','A']],\\\n",
    "                       columns=['dbSNP_ID','GENE_SYMBOL','DRUG_NAME','PMID','PHENOTYPE_CATEGORY','SIGNIFICANCE','NOTES',\\\n",
    "                               'SENTENCE','ALLELE_PharmKGB','ALLELE_23andme'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56ba8e9d",
   "metadata": {},
   "source": [
    "### Part 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "70cc6505",
   "metadata": {},
   "outputs": [],
   "source": [
    "#4. Create a tab separated file with summarized data with the following columns:\n",
    "# GENE_SYMBOL, DRUG_NAME, and a list of dbSNP_IDs separated by \";\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "375899ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TEST PART 4\n",
    "\n",
    "testDat4 = pd.read_csv(\"testData4.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b040108",
   "metadata": {},
   "source": [
    "### Part 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "65529630",
   "metadata": {},
   "outputs": [],
   "source": [
    "#5. Plot the distribution of the number of drugs associated with a gene, and the number of SNPs for a gene.\n",
    "import matplotlib.pyplot as plt\n",
    "def plot_distribution(gene_symbol, df)\n",
    "    \"\"\"\n",
    "    INPUT\n",
    "        gene_symbol: gene symbol\n",
    "        df: DataFrame that contains gene in GENE_SYMBOL column\n",
    "    OUTPUT (to change)\n",
    "        plots distribution of number of drugs associated with a gene and saves to PNG file\n",
    "        plots number of SNPs for a gene and saves to PNG file \n",
    "    \"\"\"\n",
    "#     associated columns: GENE_SYMBOL, DRUG_NAME, dbSNP_ID: list of dbSNP_IDs separated by \";\"\n",
    "    gene_rows = df.loc[df[\"GENE_SYMBOL\"] == gene_symbol]\n",
    "    associated_drugs = df.loc[:,\"DRUG_NAME\"]\n",
    "    snp_ids = df.loc[:, \"dbSNP_ID\"].split(\";\") # check dtype in df\n",
    "    snp_number = len(snp_ids)\n",
    "    # plotting\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f1972e8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# TEST PART 5\n",
    "\n",
    "#This test data is what I think you'd get from processing part 4's test data\n",
    "testDat5 = pd.DataFrame([['CYP2A6','warfarin','rs5031016'],['CYP2A6','nicotine','rs1128503;rs5031016'],['KCNMB2',\\\n",
    "             'ritodrine','rs7624046'],['ABCB1','fentanyl','rs1128503;rs1128366;rs1284621'],['ABCB1','warfarin',\\\n",
    "             'rs1128503;rs1284621'],['ABCB1','apixaban','rs1128503']],\\\n",
    "             columns=['GENE_SYMBOL','DRUG_NAME','dbSNP_IDs'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47f8439d",
   "metadata": {},
   "source": [
    "### Full Processing of Data (All functions run on data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "274c1c28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Code to run all functions on data here\n",
    "\n",
    "# Part 1\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import gzip\n",
    "variant_file = \"23andme_v5_hg19_ref.txt.gz\"\n",
    "annotation_file = \"var_drug_ann.tsv\"\n",
    "data = map_and_merge(variant_file, annotation_file)\n",
    "\n",
    "# Part 2\n",
    "data[(data.SIGNIFICANCE == \"yes\") & (data.PHENOTYPE_CATEGORY == \"Efficacy\")]\n",
    "\n",
    "# Part 3\n",
    "\n",
    "# Part 4\n",
    "\n",
    "# Part 5\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
